packages/xo-server/src/index.js

```diff
@@ -273,24 +273,34 @@ async function registerPlugin(pluginPath, pluginName) {
   })()
 
   // Supports both “normal” CommonJS and Babel's ES2015 modules.
-  const {
-    default: factory = plugin,
+  let {
+    default: instance = plugin,
     configurationSchema,
     configurationPresets,
     testSchema,
   } = plugin
 
-  // The default export can be either a factory or directly a plugin
-  // instance.
-  const instance = isFunction(factory)
-    ? factory({
-        xo: this,
-        getDataDir: () => {
-          const dir = `${this._config.datadir}/${pluginName}`
-          return ensureDir(dir).then(() => dir)
-        },
-      })
-    : factory
+  const handleFactory = factory =>
+    isFunction(factory)
+      ? factory({
+          xo: this,
+          getDataDir: () => {
+            const dir = `${this._config.datadir}/${pluginName}`
+            return ensureDir(dir).then(() => dir)
+          },
+        })
+      : factory
+  ;[
+    instance,
+    configurationSchema,
+    configurationPresets,
+    testSchema,
+  ] = await Promise.all([
+    handleFactory(instance),
+    handleFactory(configurationSchema),
+    handleFactory(configurationPresets),
+    handleFactory(testSchema),
+  ])
```

Comment:
Why these changes?

Reply:
So that the `configurationSchema` can be a factory and [access the XO instance](https://github.com/vatesfr/xen-orchestra/pull/3155/files#diff-659c438c93d2ac0a40d2e73567498d85R60) in order to list the API methods. I don't need it for `configurationPresets` and `testSchema` but I thought it would be a good thing to homogenize it.